proof/refine/AARCH64/ArchInvsLemmas_H.thy

```diff
@@ -3449,99 +327,171 @@ lemma range_cover_canonical_address:
-  apply (drule(1) pspace_distinctD')
-  apply (subst (asm) ps_clear_def)
-  apply (drule_tac x = b in orthD2)
-   apply fastforce
-  apply (drule neq_out_intv)
-   apply (simp add: mask_def add_diff_eq)
-  apply (simp add: mask_def add_diff_eq)
+lemma valid_arch_tcb_lift':
+  assumes x: "\<And>T p. \<lbrace>typ_at' T p\<rbrace> f \<lbrace>\<lambda>rv. typ_at' T p\<rbrace>"
+  shows "\<lbrace>\<lambda>s. valid_arch_tcb' tcb s\<rbrace> f \<lbrace>\<lambda>rv s. valid_arch_tcb' tcb s\<rbrace>"
+  apply (clarsimp simp add: valid_arch_tcb'_def)
+  apply (cases "atcbVCPUPtr tcb"; simp)
+   apply (wp x)+
   done
 
-lemma ptr_range_mask_range:
-  "{ptr..ptr + 2 ^ bits - 1} = mask_range ptr bits"
-  unfolding mask_def
-  by simp
-
-lemma distinct_obj_range'_not_subset:
-  "\<lbrakk> ksPSpace s a = Some obj; ksPSpace s b = Some obj'; pspace_distinct' s;
-     pspace_aligned' s; a \<noteq> b \<rbrakk>
-   \<Longrightarrow> \<not> obj_range' b obj' \<subseteq> obj_range' a obj"
-  unfolding obj_range'_def
-  apply (frule_tac x=a in pspace_alignedD')
-   apply assumption
-  apply (frule_tac x=b in pspace_alignedD')
-   apply assumption
-  apply (frule (3) rule_out_intv)
-  using is_aligned_no_overflow_mask
-  by fastforce
-
-lemma obj_range'_disjoint:
-  "\<lbrakk> ksPSpace s a = Some obj; ksPSpace s b = Some obj'; pspace_distinct' s;
-     pspace_aligned' s; a \<noteq> b \<rbrakk>
-   \<Longrightarrow> obj_range' a obj \<inter> obj_range' b obj' = {}"
-  apply (frule_tac x=a in pspace_alignedD')
-   apply assumption
-  apply (frule_tac x=b in pspace_alignedD')
-   apply assumption
-  apply (frule_tac p=a and p'=b in aligned_mask_range_cases)
-   apply assumption
-  apply (fastforce dest: distinct_obj_range'_not_subset
-                   simp: obj_range'_def)
+(* FIXME arch-split: do we want a gen_typ_at_lifts version? *)
+lemmas typ_at_lifts = typ_at_lift_tcb' typ_at_lift_ep'
+                      typ_at_lift_ntfn' typ_at_lift_cte'
+                      typ_at_lift_cte_at'
+                      typ_at_lift_page_table_at'
+                      typ_at_lift_asid_at'
+                      typ_at_lift_vcpu_at'
+                      typ_at_lift_valid_untyped'
+                      typ_at_lift_valid_cap'
+                      valid_bound_tcb_lift
+                      valid_arch_tcb_lift'
+
+lemma valid_arch_state_armKSGlobalUserVSpace:
+  "valid_arch_state' s \<Longrightarrow> canonical_address (addrFromKPPtr (armKSGlobalUserVSpace (ksArchState s)))"
+  by (simp add: valid_arch_state'_def)
+
+lemma page_table_pte_atI':
+  "\<lbrakk> page_table_at' pt_t p s; i \<le> mask (ptTranslationBits pt_t) \<rbrakk> \<Longrightarrow>
+   pte_at' (p + (i << pte_bits)) s"
+  by (simp add: page_table_at'_def)
+
+lemmas bit_simps' = pteBits_def asidHighBits_def asidPoolBits_def asid_low_bits_def
+                    asid_high_bits_def bit_simps
+
+(* FIXME arch-split: do we want to export the gen ones? *)
+lemma objBitsT_simps:
```

Comment:
Yes, I think that would be useful in proofs about those specific generic object types.

Reply:
done, but check how I did it, it's a bit height-reduced

Reply:
I think that's fine. We'll have to remember to update the number in `gen_objBitsT_simps` when we add a generic object, but I think that's reasonable.